Configure nvhpc GCC from Spack variables

pjh40 opened the pull request:
Alter the `install_components/install` script to pass the `-gcc $SPACK_CC`,
`-gpp $SPACK_CXX`, and `-g77 $SPACK_F77` flags to `makelocalrc`.  This
ensures that nvhpc is configured to use the spack gcc spec, rather than
whatever gcc is found on the path.

This commit also patches `makelocalrc` to *append* `/usr/bin` and `/bin`
to the path, rather than prepending them.  This gives precendence to
module-loaded compilers over those found in system directories.  This change
should be made in the nvhpc binary distribution.

Commit:
Configure nvhpc GCC from Spack variables

Alter the `install_components/install` script to pass the `-gcc $SPACK_CC`,
`-gpp $SPACK_CXX`, and `-g77 $SPACK_F77` flags to `makelocalrc`.  This
ensures that nvhpc is configured to use the spack gcc spec, rather than
whatever gcc is found on the path.

This commit also patches `makelocalrc` to *append* `/usr/bin` and `/bin`
to the path, rather than prepending them.  This gives precendence to
module-loaded compilers over those found in system directories.  This change
should be made in the nvhpc binary distribution.

pjh40 commented:
I'm working on the style fix, will commit soon.

Commit:
`spack style` checks now pass

samcmill commented:
It's possible to run `makelocalrc` after the installation to overwrite the localrc file.  

It would probably be more robust to just do that after calling `install`.  Something like:

```
        # Run install script
        os.system("./install")

        # Update localrc
        if spec.satisifies('%gcc'):
            os.system("/[nvhpc_prefix]/compilers/bin/makelocalrc -gcc spack_cc ... -x [nvhpc_prefix]/compilers/bin")
```

pjh40 commented:
I'll give that approach a try.

Commit:
Run `makelocalrc` rather than modifying `install`

As suggested by @samcmill, this commit runs the `..compilers/bin/makelocalrc`
script to set the specified GCC compilers, rather than modifying the
nvhpc install script.  A patch is still applied to makelocalrc to modify
the search path order.

samcmill commented:
Looks good to me, but the patch should be removed.

pjh40 commented:
Is the issue of prepending system paths to PATH fixed elsewhere?

samcmill commented:
> Is the issue of prepending system paths to PATH fixed elsewhere?

When using `-gcc`, etc. it's not an issue since it uses the specified values rather than relying on PATH.

pjh40 commented:
True, but the intent of that PATH alteration is "if the user has no PATH set, put something in that is reasonable." Instead, we have "ignore the user's PATH, and put junk in that kills their environment." This patch isn't for the package.py configuration, but for subsequent runs of the script.

I think this should be fixed upstream, and that the fix should look like this patch, but I'll take it out if you are really against it.

samcmill commented:
I agree that appending the system directories is preferable to prepending them.  I will file an internal issue for that.

However, I don't understand what subsequent runs of the script you are referring to?  makelocalrc is typically run once at install time.

pjh40 commented:
Don't you need to run it for each network host?

msimberg commented on var/spack/repos/builtin/packages/nvhpc/package.py:
```diff
@@ -89,24 +90,38 @@ class Nvhpc(Package):
+        if spec.satisfies('%gcc'):
+            prefix = Prefix(join_path(version_prefix, 'compilers'))
+            script = join_path(prefix.bin, 'makelocalrc')
+            flags = flags + ' -gcc %s -gpp %s -g77 %s -x %s' % (
+                self.compiler.cc, self.compiler.cxx,
+                self.compiler.f77, prefix.bin)
```
I gave this PR a try and it turns out that the patch (https://github.com/spack/spack/pull/24654/files#diff-cf482bcc3e73486f18429a2cf2092deee184fe53c6d55cc464e325ebb044e838) is definitely required at least on systems that have compilers in modules (I tried this on a cray system). Since `self.compiler.cc` is `cc` (another wrapper) at least in my case (likewise `cxx` and `f77`) prepending `/bin:/usr/bin` ends up having `cc` pick up the default system `g++` even if spack has otherwise ensured that the correct `g++` would be used. This makes `makelocalrc` misdetect the compilers.

The patch should also be updated to apply for different platforms.

@pjh40 do you have plans to pick this up again?

haampie commented:
Closing because #29769 is merged.
